mlflow/gateway/config.py

```diff
@@ -0,0 +1,277 @@
+import pathlib
+from enum import Enum
+import json
+import logging
+import os
+from pathlib import Path
+from pydantic import BaseModel, validator, parse_obj_as, Extra, ValidationError
+from pydantic.json import pydantic_encoder
+from typing import Optional, Union, List, Dict, Any
+import yaml
+
+from mlflow.exceptions import MlflowException
+from mlflow.gateway.constants import PROVIDERS
+from mlflow.gateway.utils import is_valid_endpoint_name, check_configuration_route_name_collisions
+
+
+_logger = logging.getLogger(__name__)
+
+
+class Provider(str, Enum):
+    CUSTOM = "custom"
+    OPENAI = "openai"
+    ANTHROPIC = "anthropic"
+    DATABRICKS_SERVING_ENDPOINT = "databricks_serving_endpoint"
+    MLFLOW = "mlflow"
+
+
+class RouteType(str, Enum):
+    CUSTOM = "custom"
+    LLM_V1_COMPLETIONS = "llm/v1/completions"
+    LLM_V1_CHAT = "llm/v1/chat"
+
+
+class OpenAIConfig(BaseModel, extra=Extra.forbid):
+    openai_api_key: Optional[str] = None
+    openai_api_type: Optional[str] = None
+    openai_api_base: Optional[str] = "https://api.openai.com/v1"
+    openai_api_version: Optional[str] = None
+    openai_organization: Optional[str] = None
+
+
+class AnthropicConfig(BaseModel, extra=Extra.forbid):
+    anthropic_api_key: Optional[str] = None
+    anthropic_api_base: Optional[str] = "https://api.anthropic.com/"
+
+
+class DatabricksConfig(BaseModel, extra=Extra.forbid):
+    databricks_api_token: Optional[str] = None
+    databricks_api_base: str
+
+
+class MLflowConfig(BaseModel, extra=Extra.forbid):
+    api_base: str
+
+
+class CustomConfig(BaseModel, extra=Extra.forbid):
+    api_key: Optional[str] = None
+    api_base: str
+    api_version: Optional[str] = None
+
+
+config_types = {
+    Provider.OPENAI: OpenAIConfig,
+    Provider.ANTHROPIC: AnthropicConfig,
+    Provider.DATABRICKS_SERVING_ENDPOINT: DatabricksConfig,
+    Provider.MLFLOW: MLflowConfig,
+    Provider.CUSTOM: CustomConfig,
+}
+
+
+class ModelInfo(BaseModel, extra=Extra.forbid):
+    name: Optional[str] = None
+    provider: Provider = Provider.CUSTOM
```

Comment:
Should the default provider be `custom`?

Reply:
No, I think we can remove it

Reply:
https://github.com/mlflow/mlflow/pull/8709